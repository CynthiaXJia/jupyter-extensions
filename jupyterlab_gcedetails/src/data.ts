/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as csstips from 'csstips';
import { stylesheet } from 'typestyle';

interface MachineType {
  name: string;
  description: string;
}

interface Instance {
  attributes: {
    framework: string;
    title: string;
    version: string;
  };
  cpuPlatform: string;
  id: number;
  image: string;
  machineType: MachineType;
  name: string;
  zone: string;
}

interface Project {
  numericProjectId: number;
  projectId: string;
}

export interface Utilization {
  cpu: number;
  memory: number;
}

interface Gpu {
  name: string;
  driver_version: string;
  cuda_version: string;
  gpu: number;
  memory: number;
  temperature: number;
}

export interface Details {
  instance: Instance;
  project: Project;
  utilization: Utilization;
  gpu: Gpu;
}

export interface Option {
  text: string;
  value: string | number;
  disabled?: boolean;
}

interface AttributeMapper {
  label: string;
  mapper: (details: Details) => string;
}

// Displayable virtual machine detail attributes
export const MAPPED_ATTRIBUTES: AttributeMapper[] = [
  { label: 'VM Name', mapper: (details: Details) => details.instance.name },
  { label: 'Project', mapper: (details: Details) => details.project.projectId },
  {
    label: 'Framework',
    mapper: (details: Details) => details.instance.attributes.framework,
  },
  {
    label: 'Machine Type',
    mapper: (details: Details) =>
      `${details.instance.machineType.description} (${details.instance.machineType.name})`,
  },
];

export const REFRESHABLE_MAPPED_ATTRIBUTES = [
  {
    label: 'CPU Utilization',
    mapper: (details: Details) => `CPU: ${details.utilization.cpu.toFixed(1)}%`,
  },
  {
    label: 'Memory Utilization',
    mapper: (details: Details) =>
      `Memory: ${details.utilization.memory.toFixed(1)}%`,
  },
  {
    label: 'GPU Utilization',
    mapper: (details: Details) => {
      if (!details.gpu.name) {
        return 'No GPUs';
      }
      return `GPU: ${details.gpu.name} - ${details.gpu.gpu.toFixed(1)}%`;
    },
  },
];

MAPPED_ATTRIBUTES.push(...REFRESHABLE_MAPPED_ATTRIBUTES);

/**
 * AI Platform Accelerator types.
 * https://cloud.google.com/ai-platform/training/docs/using-gpus#compute-engine-machine-types-with-gpu
 */
export const ACCELERATOR_TYPES: Option[] = [
  { value: '', text: 'None' },
  { value: 'NVIDIA_TESLA_K80', text: 'NVIDIA Tesla K80' },
  { value: 'NVIDIA_TESLA_P4', text: 'NVIDIA Tesla P4' },
  { value: 'NVIDIA_TESLA_P100', text: 'NVIDIA Tesla P100' },
  { value: 'NVIDIA_TESLA_T4', text: 'NVIDIA Tesla T4' },
  { value: 'NVIDIA_TESLA_V100', text: 'NVIDIA Tesla V100' },
];

/**
 * AI Platform Accelerator counts.
 * https://cloud.google.com/ai-platform/training/docs/using-gpus
 */
export const ACCELERATOR_COUNTS_1_2_4_8: Option[] = [
  { value: '1', text: '1' },
  { value: '2', text: '2' },
  { value: '4', text: '4' },
  { value: '8', text: '8' },
];

/**
 * AI Platform Machine types.
 * https://cloud.google.com/ai-platform/training/docs/machine-types#compare-machine-types
 */
export const MASTER_TYPES: Option[] = [
  { value: 'n1-standard-4', text: '4 CPUs, 15 GB RAM' },
  { value: 'n1-standard-8', text: '8 CPUs, 30 GB RAM' },
  { value: 'n1-standard-16', text: '16 CPUs, 60 GB RAM' },
  { value: 'n1-standard-32', text: '32 CPUs, 120 GB RAM' },
  { value: 'n1-standard-64', text: '64 CPUs, 240 GB RAM' },
  { value: 'n1-standard-96', text: '96 CPUs, 360 GB RAM' },

  { value: 'n1-highmem-2', text: '4 CPUs, 26 GB RAM' },
  { value: 'n1-highmem-4', text: '4 CPUs, 26 GB RAM' },
  { value: 'n1-highmem-8', text: '8 CPUs, 52 GB RAM' },
  { value: 'n1-highmem-16', text: '16 CPUs, 104 GB RAM' },
  { value: 'n1-highmem-32', text: '32 CPUs, 208 GB RAM' },
  { value: 'n1-highmem-64', text: '64 CPUs, 416 GB RAM' },
  { value: 'n1-highmem-96', text: '96 CPUs, 624 GB RAM' },

  { value: 'n1-highcpu-16', text: '16 CPUs, 14.4 GB RAM' },
  { value: 'n1-highcpu-32', text: '32 CPUs, 28.8 GB RAM' },
  { value: 'n1-highcpu-64', text: '64 CPUs, 57.6 GB RAM' },
  { value: 'n1-highcpu-96', text: '96 CPUs, 86.4 GB RAM' },
];

export interface HardwareConfiguration {
  cpu: number;
  memory: number;
}

/* CPU to Memory mappings for the Compute Engine machine types */
export interface MachineTypeConfigurations {
  base: Option;
  configurations: HardwareConfiguration[];
}

export const machineTypes: MachineTypeConfigurations[] = [
  {
    base: {
      value: 'n1-standard-',
      text: 'N1 Standard',
    },
    configurations: [
      { cpu: 4, memory: 15 },
      { cpu: 8, memory: 30 },
      { cpu: 16, memory: 60 },
      { cpu: 32, memory: 120 },
      { cpu: 64, memory: 240 },
      { cpu: 96, memory: 360 },
    ],
  },
  {
    base: {
      value: 'n1-highmem-',
      text: 'N1 High Memory',
    },
    configurations: [
      { cpu: 4, memory: 26 },
      { cpu: 8, memory: 52 },
      { cpu: 16, memory: 104 },
      { cpu: 32, memory: 208 },
      { cpu: 64, memory: 416 },
      { cpu: 96, memory: 624 },
    ],
  },
  {
    base: {
      value: 'n1-highcpu-',
      text: 'N1 High CPU',
    },
    configurations: [
      { cpu: 16, memory: 14.4 },
      { cpu: 32, memory: 28.8 },
      { cpu: 64, memory: 57.6 },
      { cpu: 96, memory: 86.4 },
    ],
  },
];

/* Class names applied to the component. */
export const STYLES = stylesheet({
  container: {
    color: 'var(--jp-ui-font-color1)',

    fontFamily: 'var(--jp-ui-font-family)',
    fontSize: 'var(--jp-ui-font-size1, 13px)',
    lineHeight: '24px',
    alignItems: 'center',
    ...csstips.horizontal,
  },
  attribute: {
    marginRight: '4px',
  },
  interactiveHover: {
    $nest: {
      '&:hover': {
        backgroundColor: '#8a8a8a',
      },
    },
  },
  dt: {
    display: 'table-cell',
    fontWeight: 'bold',
    lineHeight: '20px',
    padding: '2px',
    verticalAlign: 'top',
  },
  dd: {
    padding: '2px 2px 2px 24px',
    verticalAlign: 'top',
  },
  icon: {
    display: 'inline-block',
    height: '18px',
    marginRight: '4px',
    width: '18px',
  },
  listRow: {
    display: 'table-row',
    boxShadow: 'inset 0 -1px 0 0 var(--jp-border-color0)',
  },
});
