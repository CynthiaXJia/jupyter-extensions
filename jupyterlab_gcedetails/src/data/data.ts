/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { MachineType } from './machine_types';
import {
  Accelerator,
  nvidiaNameToEnum,
  NO_ACCELERATOR_COUNT,
} from './accelerator_types';

export interface Option {
  text: string;
  value: string | number;
  disabled?: boolean;
}

interface Instance {
  attributes: {
    framework: string;
    title: string;
    version: string;
  };
  cpuPlatform: string;
  id: number;
  image: string;
  machineType: MachineType;
  name: string;
  zone: string;
}

interface Project {
  numericProjectId: number;
  projectId: string;
}

export interface Utilization {
  cpu: number;
  memory: number;
}

interface Gpu {
  name: string;
  driver_version: string;
  cuda_version: string;
  count: string;
  gpu: number;
  memory: number;
  temperature: number;
}

export interface Details {
  instance: Instance;
  project: Project;
  utilization: Utilization;
  gpu: Gpu;
  acceleratorTypes: Accelerator[];
}

export interface HardwareConfiguration {
  machineType: MachineType;
  attachGpu: boolean;
  gpuType: string; // as Notebooks API AcceleratorType enum values
  gpuCount: string;
}

export function isEqualHardwareConfiguration(
  a: HardwareConfiguration,
  b: HardwareConfiguration
) {
  return (
    a.machineType.name === b.machineType.name &&
    a.machineType.description === b.machineType.description &&
    a.attachGpu === b.attachGpu &&
    a.gpuType === b.gpuType &&
    a.gpuCount === b.gpuCount
  );
}

export function detailsToHardwareConfiguration(
  details: Details
): HardwareConfiguration {
  const { instance, gpu } = details;

  return {
    machineType: instance.machineType,
    attachGpu: Boolean(gpu.name),
    gpuType: nvidiaNameToEnum(gpu.name),
    gpuCount: gpu.name ? gpu.count : NO_ACCELERATOR_COUNT,
  };
}

interface AttributeMapper {
  label: string;
  mapper: (details: Details) => string;
}

// Displayable virtual machine detail attributes
export const MAPPED_ATTRIBUTES: AttributeMapper[] = [
  { label: 'VM Name', mapper: (details: Details) => details.instance.name },
  { label: 'Project', mapper: (details: Details) => details.project.projectId },
  {
    label: 'Framework',
    mapper: (details: Details) => details.instance.attributes.framework,
  },
  {
    label: 'Machine Type',
    mapper: (details: Details) =>
      `${details.instance.machineType.description} (${details.instance.machineType.name})`,
  },
];

export const REFRESHABLE_MAPPED_ATTRIBUTES = [
  {
    label: 'CPU Utilization',
    mapper: (details: Details) => `CPU: ${details.utilization.cpu.toFixed(1)}%`,
  },
  {
    label: 'Memory Utilization',
    mapper: (details: Details) =>
      `Memory: ${details.utilization.memory.toFixed(1)}%`,
  },
  {
    label: 'GPU Utilization',
    mapper: (details: Details) => {
      if (!details.gpu.name) {
        return 'No GPUs';
      }
      return `GPU: ${details.gpu.name} - ${details.gpu.gpu.toFixed(1)}%`;
    },
  },
];

MAPPED_ATTRIBUTES.push(...REFRESHABLE_MAPPED_ATTRIBUTES);
