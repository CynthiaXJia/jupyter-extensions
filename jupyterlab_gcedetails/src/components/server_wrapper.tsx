import { ServerConnection } from '@jupyterlab/services';
import { Details } from '../data';

export class ServerWrapper {
  private readonly serverSettings: ServerConnection.ISettings;
  private readonly url: string;
  constructor(
    endpoint: string,
    serverSettings: ServerConnection.ISettings = ServerConnection.defaultSettings
  ) {
    this.serverSettings = serverSettings;
    this.url = this.serverSettings.baseUrl + endpoint;
  }

  async getUtilizationData(): Promise<Details> {
    const response = await ServerConnection.makeRequest(
      this.url,
      {},
      this.serverSettings
    );
    if (!response.ok) {
      throw new Error();
    }
    const data = await response.json();
    return { ...data };
  }
}
