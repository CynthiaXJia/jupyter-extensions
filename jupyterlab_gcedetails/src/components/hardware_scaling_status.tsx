import * as React from 'react';
import { stylesheet } from 'typestyle';
import LinearProgress from '@material-ui/core/LinearProgress';
import { createStyles, withStyles, Theme } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import { HardwareConfiguration } from '../data';
import { ErrorType } from './hardware_configuration_dialog';
import { NotebooksService, Instance } from '../service/notebooks_service';
import { authTokenRetrieval } from './auth_token_retrieval';
import { ServerWrapper } from './server_wrapper';

const BorderLinearProgress = withStyles((theme: Theme) =>
  createStyles({
    root: {
      width: 400,
      height: 15,
      borderRadius: 5,
    },
    colorPrimary: {
      backgroundColor:
        theme.palette.grey[theme.palette.type === 'light' ? 200 : 700],
    },
    bar: {
      backgroundColor: '#1a90ff',
      borderRadius: 5,
    },
  })
)(LinearProgress);

const STYLES = stylesheet({
  flexContainer: {
    width: 500,
    height: 300,
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
  },
  heading: {
    fontSize: '24px',
  },
  paragraph: {
    height: 60,
    width: 350,
    textAlign: 'center',
  },
});

enum Status {
  'Authorizing' = 0,
  'Stopping Instance' = 1,
  'Reshaping Instance' = 2,
  'Starting Instance' = 3,
  'Complete' = 5,
  'Error' = 6,
}

const statusInfo = [
  'Getting authorization token to reshape machine. Please complete the OAuth 2.0 authorization steps in the popup',
  'Shutting down instance for reshaping.',
  'Reshaping instance to your configuration.',
  'Restarting your instance. Your newly configured machine will be ready very shortly!',
  'Operation complete. Enjoy your newly configured instance! You may now close this dialog.',
  'An error has occured, please try again later. You may need to restart the instance manually.',
];

interface Props {
  hardwareConfiguration: HardwareConfiguration;
  notebookService: NotebooksService;
  onDialogClose: () => void;
  onCompletion: () => void;
  detailsServer: ServerWrapper;
  onError: (err, errorType: ErrorType, instanceDetails?: Instance) => void;
  showErrorPage: () => void;
}

interface State {
  status: Status;
  instanceDetails: Instance;
  reshapingError: boolean;
}

export class HardwareScalingStatus extends React.Component<Props, State> {
  private readonly NUM_RETRIES = 20;

  constructor(props: Props) {
    super(props);
    this.state = {
      status: Status.Authorizing,
      instanceDetails: null,
      reshapingError: false,
    };
  }

  private preventPageClose(event) {
    event.preventDefault();
    event.returnValue = '';
  }

  async componentDidMount() {
    const { notebookService } = this.props;
    try {
      const token = await authTokenRetrieval();
      this.setState({
        status: Status['Stopping Instance'],
      });
      notebookService.setAuthToken(token);
    } catch (err) {
      this.setState({ status: Status.Error });
      console.log(err);
    }
    window.addEventListener('beforeunload', this.preventPageClose);
  }

  async componentDidUpdate(prevProps, prevState) {
    const { status, instanceDetails, reshapingError } = this.state;
    const {
      hardwareConfiguration,
      notebookService,
      onError,
      showErrorPage,
      onCompletion,
    } = this.props;
    const { machineType, attachGpu, gpuType, gpuCount } = hardwareConfiguration;

    if (prevState.status !== status) {
      switch (status) {
        case Status['Stopping Instance']: {
          try {
            const stopResult = (await notebookService.stop()) as Instance;
            this.setState({
              status: Status['Reshaping Instance'],
              instanceDetails: stopResult,
            });
          } catch (err) {
            onError(err, ErrorType.STOP, instanceDetails);
            this.setState({ status: Status.Error });
          }
          break;
        }

        case Status['Reshaping Instance']: {
          let reshapeResult: Instance = instanceDetails;
          // Change machineType
          try {
            reshapeResult = (await notebookService.setMachineType(
              machineType.name
            )) as Instance;
          } catch (err) {
            this.setState({ reshapingError: true });
            onError(err, ErrorType.RESHAPING, reshapeResult);
          }
          // Attach GPU
          if (attachGpu) {
            try {
              reshapeResult = (await notebookService.setAccelerator(
                gpuType,
                gpuCount
              )) as Instance;
            } catch (err) {
              this.setState({ reshapingError: true });
              onError(err, ErrorType.RESHAPING, reshapeResult);
            }
          }
          // If reshaping fails, start the machine and display reshaping error message after machine has started
          this.setState({
            status: Status['Starting Instance'],
            instanceDetails: reshapeResult,
          });
          break;
        }

        case Status['Starting Instance']: {
          let startResult: Instance = null;
          try {
            startResult = (await notebookService.start()) as Instance;
          } catch (err) {
            onError(err, ErrorType.START, instanceDetails);
            this.setState({ status: Status.Error });
          }

          // Wait for server to come back up
          for (let tries = 0; tries < this.NUM_RETRIES; tries++) {
            try {
              await this.props.detailsServer.getUtilizationData();
              break;
            } catch (err) {
              if (tries === this.NUM_RETRIES - 1) {
                onError(err, ErrorType.REFRESH, startResult);
                this.setState({ status: Status.Error });
              }
              await new Promise(resolve => setTimeout(resolve, 5000));
            }
          }
          onCompletion();
          this.setState({
            status: reshapingError ? Status.Error : Status.Complete,
          });
          break;
        }

        case Status['Error']: {
          showErrorPage();
          break;
        }

        default:
          break;
      }
    }
  }

  componentWillUnmount() {
    window.removeEventListener('beforeunload', this.preventPageClose);
  }

  render() {
    const { status } = this.state;
    const progressValue = (status / 4) * 100;
    const { flexContainer, heading, paragraph } = STYLES;
    const { onDialogClose } = this.props;
    return (
      <div className={flexContainer}>
        <h3 className={heading}>{Status[status]}</h3>
        <p className={paragraph}>{statusInfo[status]}</p>
        {status === 4 || status === 5 ? (
          <Button
            variant="contained"
            color="primary"
            onClick={() => {
              onDialogClose();
            }}
          >
            Close
          </Button>
        ) : (
          <BorderLinearProgress variant="determinate" value={progressValue} />
        )}
      </div>
    );
  }
}
