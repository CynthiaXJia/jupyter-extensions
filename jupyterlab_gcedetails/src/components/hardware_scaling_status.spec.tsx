/* eslint-disable @typescript-eslint/no-empty-function */
import { shallow } from 'enzyme';
import * as React from 'react';
import { HardwareScalingStatus, Status } from './hardware_scaling_status';
import { DETAILS_RESPONSE } from '../test_helpers';
import { NotebooksService, Instance } from '../service/notebooks_service';
import { ServerWrapper } from './server_wrapper';
import { detailsToHardwareConfiguration } from '../data/data';

function immediatePromise() {
  return new Promise(r => setTimeout(r));
}

describe('HardwareScalingStatus', () => {
  const mockGetUtilizationData = jest.fn();
  const mockStopInstance = jest.fn();
  const mockSetMachineType = jest.fn();
  const mockSetAccelerator = jest.fn();
  const mockStartInstance = jest.fn();
  const mockAuthTokenRetrieval = jest.fn();
  const mockSetAuthToken = jest.fn();
  const mockOnComplete = jest.fn();
  const mockOnDialogClose = jest.fn();
  const mockServerWrapper = ({
    getUtilizationData: mockGetUtilizationData,
  } as unknown) as ServerWrapper;
  const mockNotebookService = ({
    stop: mockStopInstance,
    setMachineType: mockSetMachineType,
    setAccelerator: mockSetAccelerator,
    start: mockStartInstance,
    setAuthToken: mockSetAuthToken,
  } as unknown) as NotebooksService;
  const mockInstance: Instance = {
    name: 'Test',
  };
  const mockToken = 'mockToken';

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('Runs through reshaping flow', async () => {
    mockAuthTokenRetrieval.mockResolvedValue(mockToken);
    mockStopInstance.mockResolvedValue(mockInstance);
    mockSetMachineType.mockResolvedValue(mockInstance);
    mockSetAccelerator.mockResolvedValue(mockInstance);
    mockStartInstance.mockResolvedValue(mockInstance);
    const detailsResponse = await JSON.parse(DETAILS_RESPONSE);
    const resolveValue = Promise.resolve({ ...detailsResponse });
    mockGetUtilizationData.mockReturnValue(resolveValue);
    const hardwareConfig = detailsToHardwareConfiguration(detailsResponse);
    const hardwareScalingStatus = shallow(
      <HardwareScalingStatus
        detailsServer={mockServerWrapper}
        notebookService={mockNotebookService}
        onCompletion={mockOnComplete}
        onDialogClose={mockOnDialogClose}
        hardwareConfiguration={hardwareConfig}
        authTokenRetrieval={mockAuthTokenRetrieval}
      />
    );
    expect(hardwareScalingStatus.state('status')).toEqual(Status.Authorizing);
    expect(hardwareScalingStatus).toMatchSnapshot('Authorizing');
    expect(mockAuthTokenRetrieval).toBeCalled();
    await mockAuthTokenRetrieval();
    expect(mockNotebookService.setAuthToken).toHaveBeenCalledWith(mockToken);
    expect(hardwareScalingStatus.state('status')).toEqual(
      Status['Stopping Instance']
    );
    expect(hardwareScalingStatus).toMatchSnapshot('Stopping Instance');
    expect(mockNotebookService.stop).toBeCalled();
    await mockStopInstance();
    expect(hardwareScalingStatus.state('status')).toEqual(
      Status['Updating Machine Configuration']
    );
    expect(hardwareScalingStatus).toMatchSnapshot(
      'Updating Machine Configuration'
    );
    expect(mockNotebookService.setMachineType).toBeCalled();
    await mockSetMachineType();
    expect(hardwareScalingStatus.state('status')).toEqual(
      Status['Updating GPU Configuration']
    );
    expect(hardwareScalingStatus).toMatchSnapshot('Updating GPU Configuration');
    expect(mockNotebookService.setAccelerator).toBeCalled();
    await mockSetAccelerator();
    expect(hardwareScalingStatus.state('status')).toEqual(
      Status['Starting Instance']
    );
    expect(hardwareScalingStatus).toMatchSnapshot('Starting Instance');
    expect(mockNotebookService.start).toBeCalled();
    await mockStartInstance();
    expect(hardwareScalingStatus.state('status')).toEqual(
      Status['Refreshing Session']
    );
    expect(hardwareScalingStatus).toMatchSnapshot('Refreshing Session');
    await mockGetUtilizationData();
    await immediatePromise();
    expect(mockOnComplete).toBeCalled();
    expect(hardwareScalingStatus.state('status')).toEqual(Status.Complete);
    expect(hardwareScalingStatus).toMatchSnapshot('Complete');
  });

  it('Renders with auth error', async () => {
    const rejectedAuth = Promise.reject('No token');
    mockAuthTokenRetrieval.mockReturnValue(rejectedAuth);
    const detailsResponse = JSON.parse(DETAILS_RESPONSE);
    const hardwareConfig = detailsToHardwareConfiguration(detailsResponse);
    const hardwareScalingStatus = shallow(
      <HardwareScalingStatus
        detailsServer={mockServerWrapper}
        notebookService={mockNotebookService}
        onCompletion={mockOnComplete}
        onDialogClose={mockOnDialogClose}
        hardwareConfiguration={hardwareConfig}
        authTokenRetrieval={mockAuthTokenRetrieval}
      />
    );
    await rejectedAuth.catch(() => {});
    expect(hardwareScalingStatus.state('status')).toEqual(Status.Error);
    expect(hardwareScalingStatus).toMatchSnapshot('Auth Error');
  });

  it('Renders with stop error', async () => {
    mockAuthTokenRetrieval.mockResolvedValue(mockToken);
    const rejectedStopOperation = Promise.reject('Stop operation failed');
    mockStopInstance.mockReturnValue(rejectedStopOperation);
    const detailsResponse = JSON.parse(DETAILS_RESPONSE);
    const hardwareConfig = detailsToHardwareConfiguration(detailsResponse);
    const hardwareScalingStatus = shallow(
      <HardwareScalingStatus
        detailsServer={mockServerWrapper}
        notebookService={mockNotebookService}
        onCompletion={mockOnComplete}
        onDialogClose={mockOnDialogClose}
        hardwareConfiguration={hardwareConfig}
        authTokenRetrieval={mockAuthTokenRetrieval}
      />
    );
    await mockAuthTokenRetrieval();
    await rejectedStopOperation.catch(() => {});
    expect(hardwareScalingStatus.state('status')).toEqual(Status.Error);
    expect(hardwareScalingStatus).toMatchSnapshot('Stop instance error');
  });

  it('Renders with operation error', async () => {
    const mockGetUtilizationData = jest.fn();
    const mockStopInstance = jest.fn();
    const rejectedSetMachine = Promise.reject('Reshape operation failed');
    const rejectedSetAccelerator = Promise.reject('Reshape operation failed');
    mockSetMachineType.mockReturnValue(rejectedSetMachine);
    mockSetAccelerator.mockReturnValue(rejectedSetAccelerator);
    mockStartInstance.mockResolvedValue(mockInstance);
    const detailsResponse = JSON.parse(DETAILS_RESPONSE);
    const resolveValue = Promise.resolve({ ...detailsResponse });
    mockGetUtilizationData.mockReturnValue(resolveValue);
    const hardwareConfig = detailsToHardwareConfiguration(detailsResponse);
    const hardwareScalingStatus = shallow(
      <HardwareScalingStatus
        detailsServer={mockServerWrapper}
        notebookService={mockNotebookService}
        onCompletion={mockOnComplete}
        onDialogClose={mockOnDialogClose}
        hardwareConfiguration={hardwareConfig}
        authTokenRetrieval={mockAuthTokenRetrieval}
      />
    );
    await mockAuthTokenRetrieval();
    await mockStopInstance();
    await rejectedSetMachine.catch(() => {});
    await rejectedSetAccelerator.catch(() => {});
    await mockStartInstance();
    await immediatePromise();
    expect(hardwareScalingStatus.state('status')).toEqual(Status.Error);
    expect(hardwareScalingStatus).toMatchSnapshot('Operation error');
  });

  it('Renders with restart error', async () => {
    const mockGetUtilizationData = jest.fn();
    const mockStopInstance = jest.fn();
    const rejectedSetMachine = Promise.reject('Reshape operation failed');
    const rejectedSetAccelerator = Promise.reject('Reshape operation failed');
    const rejectedStartOperation = Promise.reject('Restart operation failed');
    mockSetMachineType.mockReturnValue(rejectedSetMachine);
    mockSetAccelerator.mockReturnValue(rejectedSetAccelerator);
    mockStartInstance.mockRejectedValue(rejectedStartOperation);
    const detailsResponse = JSON.parse(DETAILS_RESPONSE);
    const resolveValue = Promise.resolve({ ...detailsResponse });
    mockGetUtilizationData.mockReturnValue(resolveValue);
    const hardwareConfig = detailsToHardwareConfiguration(detailsResponse);
    const hardwareScalingStatus = shallow(
      <HardwareScalingStatus
        detailsServer={mockServerWrapper}
        notebookService={mockNotebookService}
        onCompletion={mockOnComplete}
        onDialogClose={mockOnDialogClose}
        hardwareConfiguration={hardwareConfig}
        authTokenRetrieval={mockAuthTokenRetrieval}
      />
    );
    await mockAuthTokenRetrieval();
    await mockStopInstance();
    await rejectedSetMachine.catch(() => {});
    await rejectedSetAccelerator.catch(() => {});
    await rejectedStartOperation.catch(() => {});
    expect(hardwareScalingStatus.state('status')).toEqual(Status.Error);
    expect(hardwareScalingStatus).toMatchSnapshot('Restart error');
  });
});
