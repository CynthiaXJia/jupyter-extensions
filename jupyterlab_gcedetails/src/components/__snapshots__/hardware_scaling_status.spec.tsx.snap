// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HardwareScalingStatus Renders with auth error: Auth Error 1`] = `
<ErrorPage
  error={null}
  instanceDetails={null}
  onDialogClose={[MockFunction]}
/>
`;

exports[`HardwareScalingStatus Renders with operation error: Operation error 1`] = `
<ErrorPage
  error={
    Object {
      "errorType": "Reshape",
      "errorValue": "Reshape operation failed",
    }
  }
  instanceDetails={
    Object {
      "name": "Test",
    }
  }
  onDialogClose={[MockFunction]}
/>
`;

exports[`HardwareScalingStatus Renders with restart error: Restart error 1`] = `
<ErrorPage
  error={
    Object {
      "errorType": "Start",
      "errorValue": Promise {},
    }
  }
  onDialogClose={[MockFunction]}
/>
`;

exports[`HardwareScalingStatus Renders with stop error: Stop instance error 1`] = `
<ErrorPage
  error={
    Object {
      "errorType": "Stop",
      "errorValue": "Stop operation failed",
    }
  }
  instanceDetails={null}
  onDialogClose={[MockFunction]}
/>
`;

exports[`HardwareScalingStatus Runs through reshaping flow: Authorizing 1`] = `
<div
  className="containerPadding_f1ms73ho"
>
  <div
    className="containerSize_fjiu9nv"
  >
    <p
      className="heading_f1xburhd"
    >
      Authorizing
    </p>
    <p
      className="paragraph_fssm9cv"
    >
      Please complete the OAuth 2.0 authorization steps in the popup
    </p>
  </div>
  <WithStyles(WithStyles(ForwardRef(LinearProgress)))
    value={0}
    variant="determinate"
  />
</div>
`;

exports[`HardwareScalingStatus Runs through reshaping flow: Complete 1`] = `
<div
  className="containerPadding_f1ms73ho"
>
  <div
    className="containerSize_fjiu9nv"
  >
    <p
      className="heading_f1xburhd"
    >
      Complete
    </p>
    <p
      className="paragraph_fssm9cv"
    >
      Operation complete. Enjoy your newly configured instance!
    </p>
  </div>
  <ActionBar
    onPrimaryClick={[MockFunction]}
    primaryLabel="Close"
  />
</div>
`;

exports[`HardwareScalingStatus Runs through reshaping flow: Refreshing Session 1`] = `
<div
  className="containerPadding_f1ms73ho"
>
  <div
    className="containerSize_fjiu9nv"
  >
    <p
      className="heading_f1xburhd"
    >
      Refreshing Session
    </p>
    <p
      className="paragraph_fssm9cv"
    >
      Refreshing your JupyterLab session to reflect your new configuration.
    </p>
  </div>
  <WithStyles(WithStyles(ForwardRef(LinearProgress)))
    value={83.33333333333334}
    variant="determinate"
  />
</div>
`;

exports[`HardwareScalingStatus Runs through reshaping flow: Starting Instance 1`] = `
<div
  className="containerPadding_f1ms73ho"
>
  <div
    className="containerSize_fjiu9nv"
  >
    <p
      className="heading_f1xburhd"
    >
      Starting Instance
    </p>
    <p
      className="paragraph_fssm9cv"
    >
      Restarting your instance.
    </p>
  </div>
  <WithStyles(WithStyles(ForwardRef(LinearProgress)))
    value={66.66666666666666}
    variant="determinate"
  />
</div>
`;

exports[`HardwareScalingStatus Runs through reshaping flow: Stopping Instance 1`] = `
<div
  className="containerPadding_f1ms73ho"
>
  <div
    className="containerSize_fjiu9nv"
  >
    <p
      className="heading_f1xburhd"
    >
      Stopping Instance
    </p>
    <p
      className="paragraph_fssm9cv"
    >
      Shutting down instance for reshaping.
    </p>
  </div>
  <WithStyles(WithStyles(ForwardRef(LinearProgress)))
    value={16.666666666666664}
    variant="determinate"
  />
</div>
`;

exports[`HardwareScalingStatus Runs through reshaping flow: Updating GPU Configuration 1`] = `
<div
  className="containerPadding_f1ms73ho"
>
  <div
    className="containerSize_fjiu9nv"
  >
    <p
      className="heading_f1xburhd"
    >
      Updating GPU Configuration
    </p>
    <p
      className="paragraph_fssm9cv"
    >
      Reshaping GPU configuration to match your selection.
    </p>
  </div>
  <WithStyles(WithStyles(ForwardRef(LinearProgress)))
    value={50}
    variant="determinate"
  />
</div>
`;

exports[`HardwareScalingStatus Runs through reshaping flow: Updating Machine Configuration 1`] = `
<div
  className="containerPadding_f1ms73ho"
>
  <div
    className="containerSize_fjiu9nv"
  >
    <p
      className="heading_f1xburhd"
    >
      Updating Machine Configuration
    </p>
    <p
      className="paragraph_fssm9cv"
    >
      Reshaping machine configuration to match your selection.
    </p>
  </div>
  <WithStyles(WithStyles(ForwardRef(LinearProgress)))
    value={33.33333333333333}
    variant="determinate"
  />
</div>
`;
